server/src/main.rs

```diff
@@ -104,21 +119,40 @@ fn create_response(req: Request<Body>) -> impl Future<Item = Response<Body>, Err
     }
 }
 
+fn extract_signature(
+    headers: HeaderMap<HeaderValue>,
+    query_opt: Option<&str>,
```

Comment:
This is a weird type signature, first argument is strongly typed, second one is an opaque string. It'd be better to accept `Option<url::form_urlencoded::Parse>` or even `Option<HashMap<&str, &str>>` here instead.